skyportal/facility_apis/lt.py

```diff
@@ -0,0 +1,855 @@
+                "type": "number",
+                "default": 1.2,
+            },
+            "sky_brightness": {
+                "title": "Maximum allowable Sky Brightness, Dark + X magnitudes [arcsec] (0-5)",
+                "type": "number",
+                "default": 2.0,
+            },
+            "photometric": {
+                "title": "Does this observation require photometric conditions?",
+                "type": "boolean",
+            },
+            "binning": {"type": "string", "enum": ["1x1", "2x2"], "default": "1x1"},
+        },
+        "required": [
+            "observation_type",
+            "exposure_type",
+            "priority",
+            "start_date",
+            "end_date",
+            "maximum_airmass",
+            "maximum_seeing",
+            "binning",
+        ],
+    }
+
+    ui_json_schema = {}
+
+
+class SPRATAPI(LTAPI):
+
+    """An interface to LT SPRAT operations."""
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Submit a follow-up request to LT's SPRAT.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+        """
+
+        from ..models import DBSession, FacilityTransaction
+
+        altdata = request.allocation.load_altdata()
+        if not altdata:
+            return
+
+        ltreq = SPRATRequest()
+        observation_payload = ltreq._build_prolog()
+        ltreq._build_project(observation_payload, request)
+        ltreq._build_inst_schedule(observation_payload, request)
+
+        headers = {
+            'Username': altdata["username"],
+            'Password': altdata["password"],
+        }
+        url = '{0}://{1}:{2}/node_agent2/node_agent?wsdl'.format(
+            'http', cfg['app.lt_host'], cfg['app.lt_port']
+        )
+        client = Client(url=url, headers=headers)
+        full_payload = etree.tostring(
+            observation_payload, encoding="unicode", pretty_print=True
+        )
+        # Send payload, and receive response string, removing the encoding tag which causes issue with lxml parsing
+        response = client.service.handle_rtml(full_payload).replace(
+            'encoding="ISO-8859-1"', ''
+        )
+        response_rtml = etree.fromstring(response)
+        mode = response_rtml.get('mode')
+
+        if mode == 'confirm':
+            request.status = 'submitted'
+        else:
+            request.status = f'rejected: {mode}'
+
+        transaction = FacilityTransaction(
+            request=http.serialize_requests_request_xml(full_payload),
+            response=http.serialize_requests_response_xml(response),
+            followup_request=request,
+            initiator_id=request.last_modified_by_id,
+        )
+
+        DBSession().add(transaction)
+
+    _instrument_configs = {}
+
+    _observation_types = ['blue', 'red']
```

Comment:
Generally people will want to expose both of these simultaneously, right? It seems like the current implementation forces you to pick one or the other at a time to request

Reply:
My understanding from the existing interfaces that you can only choose one or the other. Let's ask Dan on slack.

Reply:
Ok confirmed by Dan.